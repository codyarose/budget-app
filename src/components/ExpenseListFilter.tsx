import React, { FC, ChangeEvent, useState } from 'react'
import { useSelector, useDispatch } from 'react-redux'
import { RootState } from '../app/store'
import {
	setTextFilter,
	sortByDate,
	sortByAmount,
	setStartDate,
	setEndDate,
} from '../features/filter/filterSlice'
import { DateRangePicker } from 'react-dates'

const ExpenseListFilter: FC = () => {
	const { text, sortBy, startDate, endDate } = useSelector(
		(state: RootState) => state.filter,
	)
	const dispatch = useDispatch()

	const [calendarFocused, setCalendarFocused] = useState<
		'startDate' | 'endDate' | null
	>(null)

	const handleTextFilterChange = (e: ChangeEvent<HTMLInputElement>) => {
		dispatch(setTextFilter(e.target.value))
	}

	const handleSortByChange = (e: ChangeEvent<HTMLSelectElement>) => {
		const { value } = e.target
		value === 'date'
			? dispatch(sortByDate())
			: value === 'amount' && dispatch(sortByAmount())
	}

	const handleDatesChange = ({ startDate, endDate }: any) => {
		dispatch(setStartDate(startDate))
		dispatch(setEndDate(endDate))
	}

	const handleFocusChange = (
		calendarFocused: 'startDate' | 'endDate' | null,
	) => {
		setCalendarFocused(calendarFocused)
	}

	return (
		<div>
			<input
				type="text"
				name="textFilter"
				value={text}
				onChange={handleTextFilterChange}
			/>
			<select
				name="sortFilter"
				value={sortBy}
				onChange={handleSortByChange}
			>
				<option value="date">Date</option>
				<option value="amount">Amount</option>
			</select>
			<DateRangePicker
				startDate={startDate}
				startDateId="startDateId"
				endDate={endDate}
				endDateId="endDateId"
				onDatesChange={handleDatesChange}
				focusedInput={calendarFocused}
				onFocusChange={handleFocusChange}
				numberOfMonths={1}
				isOutsideRange={() => false}
				showClearDates={true}
			/>
		</div>
	)
}

export default ExpenseListFilter
