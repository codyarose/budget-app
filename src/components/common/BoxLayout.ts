import styled from 'styled-components'

const Layout = styled.div`
	background: ${({ theme }) => theme.colors.white};
	height: 100vh;
	display: flex;
	justify-content: center;
	align-items: center;
`

const Box = styled.div`
	position: relative;
	background-color: ${({ theme }) => theme.colors.white};
	border: 3px solid ${({ theme }) => theme.colors.black};
	text-align: center;
	padding: ${({ theme }) => `${theme.spacing.md} ${theme.spacing.lg}`};
	box-shadow: ${({ theme }) =>
		`${theme.spacing.xs} ${theme.spacing.xs} 0 currentColor`};
`

const BoxLayout = {
	Layout,
	Box,
}

export default BoxLayout
