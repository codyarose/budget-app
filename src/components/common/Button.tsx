import React, { FC } from 'react'
import styled, { css } from 'styled-components'
import { transitions } from 'polished'

interface Props {
	onClick?: () => void
	className?: string
	title?: string
}

const Button: FC<Props> = ({ onClick, children, className, title }) => {
	return (
		<StyledButton onClick={onClick} className={className} title={title}>
			{children}
		</StyledButton>
	)
}

export default Button

export const buttonStyles = css`
	background-color: ${({ theme }) => theme.colors.bg};
	color: ${({ theme }) => theme.colors.fg};
	border: 2px solid currentColor;
	padding: ${({ theme }) => `${theme.spacing.xs} ${theme.spacing.sm}`};
	font-size: 1rem;
	box-shadow: 3px 3px 0 currentColor;
	cursor: pointer;
	${transitions(
		'box-shadow 0.1s ease-in-out',
		'transform 0.1s ease-in-out',
		'background-color 0.2s ease-in-out',
		'color 0.2s ease-in-out',
	)}
	&:hover,
	&:focus {
		box-shadow: 5px 5px 0 currentColor;
		outline: none;
	}
	&:active {
		transform: translate(1px, 1px);
		box-shadow: 4px 4px 0 currentColor;
	}
`

const StyledButton = styled.button`
	${buttonStyles}
`
