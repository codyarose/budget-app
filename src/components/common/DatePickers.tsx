import React, { FC } from 'react'
import styled, { css } from 'styled-components'
import { rgba } from 'polished'
import {
	DateRangePicker,
	DateRangePickerShape,
	SingleDatePicker,
	SingleDatePickerShape,
} from 'react-dates'

export const StyledDateRangePicker: FC<DateRangePickerShape> = (props) => {
	return (
		<DateRangeWrapper>
			<DateRangePicker {...props} />
		</DateRangeWrapper>
	)
}

export const StyledSingleDatePicker: FC<SingleDatePickerShape> = (props) => {
	return (
		<SingleDateWrapper>
			<SingleDatePicker {...props} />
		</SingleDateWrapper>
	)
}
StyledSingleDatePicker.displayName = 'StyledSingleDatePicker'

const commonStyles = css`
	--background: ${({ theme }) => theme.colors.bg};
	--foreground: ${({ theme }) => theme.colors.fg};
	color: var(--foreground);
	transition: color 0.2s ease-in-out;

	.DayPicker__horizontal,
	.CalendarMonth,
	.CalendarMonthGrid {
		background: inherit;
	}
	.DateInput_input,
	.DayPicker_weekHeader,
	.CalendarMonth_caption {
		color: inherit;
	}
	.CalendarDay__default,
	.DayPickerNavigation_button__default {
		background: inherit;
		color: inherit;
	}

	.DateInput {
		height: 100%;
		background-color: transparent;
	}
	.DateInput_input {
		height: 100%;
		background-color: inherit;
		&__focused {
			border-color: var(--foreground);
		}
	}
	.DayPicker {
		&__withBorder {
			box-shadow: ${({ theme }) =>
				`0 2px 6px ${rgba(theme.colors.fg, 0.05)}, 0 0 0 1px ${rgba(
					theme.colors.fg,
					0.07,
				)}`};
		}
		&__horizontal {
		}
		&_weekHeader {
		}
	}
	.CalendarMonth,
	.CalendarMonthGrid {
	}
	.CalendarMonth_caption {
	}
	.CalendarDay {
		&__default {
		}
		&__selected {
			background: var(--foreground);
			border-color: var(--foreground);
			color: var(--background);
		}
		&__selected_span {
			background: ${({ theme }) => rgba(theme.colors.fg, 0.5)};
			border-color: ${({ theme }) => rgba(theme.colors.fg, 0.25)};
			color: var(--background);
			&:hover {
				background: ${({ theme }) => rgba(theme.colors.fg, 0.6)};
			}
		}
	}
	.DayPickerKeyboardShortcuts_show__bottomRight::before {
		border-right-color: var(--foreground);
	}
	.DayPickerKeyboardShortcuts_showSpan {
		color: var(--background);
	}
	.DayPickerNavigation_button__default {
		border: 1px solid ${({ theme }) => rgba(theme.colors.fg, 0.5)};
	}
	.DayPickerNavigation_svg__horizontal {
		fill: var(--foreground);
	}
`

const SingleDateWrapper = styled.div`
	.SingleDatePicker {
		width: 100%;
	}
	.SingleDatePickerInput {
		width: 100%;
		background-color: transparent;

		&__withBorder {
			border: none;
			border-bottom: 2px solid var(--foreground);
			transition: border-color 0.2s ease-in-out;
		}
	}
	${commonStyles}
	.DateInput {
		width: 100%;
	}
`

const DateRangeWrapper = styled.div`
	.DateRangePicker {
		height: 100%;
		& > div {
			height: 100%;
		}
		&_picker {
			background-color: var(--background);
		}
	}
	.DateRangePickerInput {
		height: 100%;
		background-color: transparent;
		&__withBorder {
			border: none;
			border-bottom: 2px solid var(--foreground);
			transition: border-color 0.2s ease-in-out;
		}
		&_arrow_svg,
		&_clearDates_svg {
			fill: var(--foreground);
			transition: fill 0.22s ease-in-out;
		}
	}

	${commonStyles}
`
