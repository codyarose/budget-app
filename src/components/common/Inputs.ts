import styled, { css } from 'styled-components'
import { transitions } from 'polished'
const baseStyles = css`
	background-color: transparent;
	color: ${({ theme }) => theme.colors.fg};
	font-size: 1rem;
	border: none;
	border-bottom: 2px solid currentColor;
	padding: ${({ theme }) => `${theme.spacing.sm}`};
	${transitions(['color', 'border-color'], '0.2s ease-in-out')}
`

export const TextInput = styled.input`
	${baseStyles}
	&::placeholder {
		color: ${({ theme }) => theme.colors.fg};
		transition: color 0.2s ease-in-out;
	}
`

export const SelectInput = styled.select`
	${baseStyles}
`
