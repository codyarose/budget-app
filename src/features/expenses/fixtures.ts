export default [
	{
		id: '123abc',
		description: 'Title',
		note: '',
		amount: 140500,
		createdAt: 1598763926001,
	},
	{
		id: '456def',
		description: 'Blah',
		note: 'This is a note',
		amount: 2650,
		createdAt: 1598763926009,
	},
	{
		id: '789ghi',
		description: 'Another one',
		note: 'Text goes here',
		amount: 340098,
		createdAt: 1598763926345,
	},
]
