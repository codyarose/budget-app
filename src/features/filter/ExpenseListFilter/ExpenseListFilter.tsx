import React, { FC, ChangeEvent, useState } from 'react'
import { useSelector, useDispatch } from 'react-redux'
import {
	setTextFilter,
	sortByDate,
	sortByAmount,
	setStartDate,
	setEndDate,
	selectText,
	selectSortBy,
	selectStartDate,
	selectEndDate,
} from '../filterSlice'
import { DateRangePicker } from 'react-dates'
import moment from 'moment'

const ExpenseListFilter: FC = () => {
	const textFilter = useSelector(selectText)
	const sortByFilter = useSelector(selectSortBy)
	const startDateFilter = useSelector(selectStartDate)
	const endDateFilter = useSelector(selectEndDate)
	const dispatch = useDispatch()

	const momentStartDate = moment(startDateFilter)
	const momentEndDate = moment(endDateFilter)

	const [calendarFocused, setCalendarFocused] = useState<
		'startDate' | 'endDate' | null
	>(null)

	const handleTextFilterChange = (e: ChangeEvent<HTMLInputElement>) => {
		dispatch(setTextFilter(e.target.value))
	}

	const handleSortByChange = (e: ChangeEvent<HTMLSelectElement>) => {
		const { value } = e.target
		value === 'date'
			? dispatch(sortByDate())
			: value === 'amount' && dispatch(sortByAmount())
	}

	const handleDatesChange = ({ startDate, endDate }: any) => {
		dispatch(setStartDate(startDate))
		dispatch(setEndDate(endDate))
	}

	const handleFocusChange = (
		calendarFocused: 'startDate' | 'endDate' | null,
	) => {
		setCalendarFocused(calendarFocused)
	}

	return (
		<div>
			<input
				type="text"
				name="textFilter"
				value={textFilter}
				onChange={handleTextFilterChange}
			/>
			<select
				name="sortFilter"
				value={sortByFilter}
				onChange={handleSortByChange}
			>
				<option value="date">Date</option>
				<option value="amount">Amount</option>
			</select>
			<DateRangePicker
				startDate={momentStartDate}
				startDateId="startDateId"
				endDate={momentEndDate}
				endDateId="endDateId"
				onDatesChange={handleDatesChange}
				focusedInput={calendarFocused}
				onFocusChange={handleFocusChange}
				numberOfMonths={1}
				isOutsideRange={() => false}
				showClearDates={true}
			/>
		</div>
	)
}

export default ExpenseListFilter
