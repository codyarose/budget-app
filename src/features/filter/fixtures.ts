import moment from 'moment'

const filters = {
	text: '',
	sortBy: 'date',
	startDate: moment(0),
	endDate: moment(0).add(3, 'days'),
}

const altFilters = {
	text: 'bills',
	sortBy: 'amount',
	startDate: moment(0),
	endDate: moment(0).add(10, 'days'),
}

export { filters, altFilters }
