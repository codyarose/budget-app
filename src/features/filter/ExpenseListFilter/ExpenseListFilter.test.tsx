import React from 'react'
import { shallow, ShallowWrapper } from 'enzyme'
import * as redux from 'react-redux'
import moment from 'moment'

import ExpenseListFilter from './index'
import {
	selectText,
	selectSortBy,
	selectStartDate,
	selectEndDate,
	setTextFilter,
	sortByDate,
	sortByAmount,
	setStartDate,
	setEndDate,
} from '../filterSlice'
import { filters, altFilters } from '../fixtures'
import { TextInput, SelectInput } from '../../../components/common/Inputs'
import { StyledDateRangePicker } from '../../../components/common/DatePickers'

jest.mock('moment', () => {
	const moment = jest.requireActual('moment')
	const mockMoment = (timestamp = 0) => {
		return moment(timestamp)
	}
	return mockMoment
})

jest.mock('react-redux', () => ({
	useDispatch: jest.fn(),
	useSelector: jest.fn((fn) => fn()),
}))

jest.mock('../filterSlice')

const useDispatchSpy = jest.spyOn(redux, 'useDispatch')
const mockDispatchFn = jest.fn()
useDispatchSpy.mockReturnValue(mockDispatchFn)

let wrapper: ShallowWrapper
beforeEach(() => {
	wrapper = shallow(<ExpenseListFilter />)
})

describe('ExpenseListFilter', () => {
	it('should render correctly', () => {
		;(selectText as jest.Mock).mockReturnValue(filters.text)
		;(selectSortBy as jest.Mock).mockReturnValue(filters.sortBy)
		;(selectStartDate as jest.Mock).mockReturnValue(filters.startDate)
		;(selectEndDate as jest.Mock).mockReturnValue(filters.endDate)
		wrapper.setProps({})
		expect(wrapper).toMatchSnapshot()
	})

	it('should render with alt data correctly', () => {
		;(selectText as jest.Mock).mockReturnValueOnce(altFilters.text)
		;(selectSortBy as jest.Mock).mockReturnValueOnce(altFilters.sortBy)
		;(selectStartDate as jest.Mock).mockReturnValueOnce(
			altFilters.startDate,
		)
		;(selectEndDate as jest.Mock).mockReturnValueOnce(altFilters.endDate)
		wrapper.setProps({})
		expect(wrapper).toMatchSnapshot()
	})

	it('should handle text change', () => {
		const value = 'new text'

		wrapper.find(TextInput).simulate('change', { target: { value } })

		expect(setTextFilter).toHaveBeenLastCalledWith(value)
	})

	it('should handle sort by date', () => {
		const value = 'date'

		;(selectSortBy as jest.Mock).mockReturnValueOnce(altFilters.sortBy)
		wrapper.setProps({})
		expect(wrapper.find(SelectInput).at(0).prop('value')).toBe(
			altFilters.sortBy,
		)

		wrapper
			.find(SelectInput)
			.at(0)
			.simulate('change', { target: { value } })
		wrapper.setProps({})
		expect(wrapper.find(SelectInput).at(0).prop('value')).toBe(value)
		expect(sortByDate).toHaveBeenCalled()
	})

	it('should handle sort by amount', () => {
		const value = 'amount'

		wrapper
			.find(SelectInput)
			.at(0)
			.simulate('change', { target: { value } })

		expect(sortByAmount).toHaveBeenCalled()
	})

	it('should handle date changes', () => {
		const startDate = moment(0).add(6, 'days')
		const endDate = moment(0).add(12, 'days')

		wrapper.find(StyledDateRangePicker).prop('onDatesChange')({
			startDate,
			endDate,
		})

		expect(setStartDate).toHaveBeenLastCalledWith(startDate)
		expect(setEndDate).toHaveBeenLastCalledWith(endDate)
	})

	it('should handle date focus changes', () => {
		const value = 'endDate'

		wrapper.find(StyledDateRangePicker).prop('onFocusChange')(value)
		expect(wrapper.find(StyledDateRangePicker).prop('focusedInput')).toBe(
			value,
		)
	})
})
