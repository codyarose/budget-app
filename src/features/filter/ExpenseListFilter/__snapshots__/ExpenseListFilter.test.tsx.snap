// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpenseListFilter should render correctly 1`] = `
<styled.div>
  <styled.input
    name="textFilter"
    onChange={[Function]}
    placeholder="Search expenses"
    type="text"
    value=""
  />
  <styled.select
    name="sortFilter"
    onChange={[Function]}
    value="date"
  >
    <option
      value="date"
    >
      Date
    </option>
    <option
      value="amount"
    >
      Amount
    </option>
  </styled.select>
  <StyledDateRangePicker
    endDate={"1970-01-04T00:00:00.000Z"}
    endDateId="endDateId"
    focusedInput={null}
    isOutsideRange={[Function]}
    numberOfMonths={1}
    onDatesChange={[Function]}
    onFocusChange={[Function]}
    showClearDates={true}
    startDate={"1970-01-01T00:00:00.000Z"}
    startDateId="startDateId"
  />
</styled.div>
`;

exports[`ExpenseListFilter should render with alt data correctly 1`] = `
<styled.div>
  <styled.input
    name="textFilter"
    onChange={[Function]}
    placeholder="Search expenses"
    type="text"
    value="bills"
  />
  <styled.select
    name="sortFilter"
    onChange={[Function]}
    value="amount"
  >
    <option
      value="date"
    >
      Date
    </option>
    <option
      value="amount"
    >
      Amount
    </option>
  </styled.select>
  <StyledDateRangePicker
    endDate={"1970-01-11T00:00:00.000Z"}
    endDateId="endDateId"
    focusedInput={null}
    isOutsideRange={[Function]}
    numberOfMonths={1}
    onDatesChange={[Function]}
    onFocusChange={[Function]}
    showClearDates={true}
    startDate={"1970-01-01T00:00:00.000Z"}
    startDateId="startDateId"
  />
</styled.div>
`;
